dev/com.ibm.ws.microprofile.metrics.2.3/src/com/ibm/ws/microprofile/metrics23/impl/SimpleTimerImpl.java

```diff
@@ -0,0 +1,130 @@
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.microprofile.metrics.Counter;
+import org.eclipse.microprofile.metrics.Meter;
+import org.eclipse.microprofile.metrics.SimpleTimer;
+
+import com.ibm.ws.microprofile.metrics.impl.Clock;
+import com.ibm.ws.microprofile.metrics.impl.CounterImpl;
+
+/**
+ * A timer metric which aggregates timing durations and provides duration statistics, plus
+ * throughput statistics via {@link Meter}.
+ */
+public class SimpleTimerImpl implements SimpleTimer {
+    /**
+     * A timing context.
+     *
+     * @see SimpleTimerImpl#time()
+     */
+    public static class Context implements SimpleTimer.Context {
+        private final SimpleTimerImpl simpleTimer;
+        private final Clock clock;
+        private final long startTime;
+
+        private Context(SimpleTimerImpl simpleTimer, Clock clock) {
+            this.simpleTimer = simpleTimer;
+            this.clock = clock;
+            this.startTime = clock.getTick();
+        }
+
+        /**
+         * Updates the timer with the difference between current and start time. Call to this method will
+         * not reset the start time. Multiple calls result in multiple updates.
+         *
+         * @return the elapsed time in nanoseconds
+         */
+        @Override
+        public long stop() {
+            final long elapsed = clock.getTick() - startTime;
+            simpleTimer.update(elapsed, TimeUnit.NANOSECONDS);
+            return elapsed;
+        }
+
+        /** Equivalent to calling {@link #stop()}. */
+        @Override
+        public void close() {
+            stop();
+        }
+    }
+
+    private final Counter count;
+    private long elapsedTime;
+    private final Clock clock;
+
+    public SimpleTimerImpl() {
+        this(Clock.defaultClock());
+    }
+
+    public SimpleTimerImpl(Clock clock) {
+        this.clock = clock;
+        this.count = new CounterImpl();
+        this.elapsedTime = 0;
+    }
+
+    /**
+     * Adds a recorded duration.
+     *
+     * @param duration the length of the duration
+     * @param unit     the scale unit of {@code duration}
+     */
+    @Override
+    public void update(long duration, TimeUnit unit) {
+        update(unit.toNanos(duration));
+    }
+
+    @Override
+    public <T> T time(Callable<T> event) throws Exception {
+        final long startTime = clock.getTick();
+        try {
+            return event.call();
+        } finally {
+            update(clock.getTick() - startTime);
+        }
+    }
+
+    @Override
+    public void time(Runnable event) {
+        final long startTime = clock.getTick();
+        try {
+            event.run();
+        } finally {
+            update(clock.getTick() - startTime);
+        }
+    }
+
+    @Override
+    public Context time() {
+        return new Context(this, clock);
+    }
+
+    private void update(long duration) {
+        elapsedTime += duration;
```

Comment:
Do we need synchronization when using SimpleTimer in multiple threads?

Reply:
synchronized added